refactor: call args as an array instead of a slice

benesjan opened the pull request:
Usage of slices is now discouraged and for this reason I am dropping their use from calls.

This made me drop the `hash_args` function and rename `hash_args_array` to just `hash_args` as the original was no longer in use. Also `poseidon2_hash_with_separator_slice` got dropped due to lack of use.

benesjan commented:
* **#18417** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18417?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18417?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18413** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18413?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18396** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18396?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18398** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18398?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18274** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18274?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18265** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18265?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17963** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17963?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18126** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18126?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17792** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/test/helpers/authwit.nr:
```diff
@@ -10,11 +10,11 @@ use protocol_types::{
 };
 
-pub unconstrained fn add_private_authwit_from_call<T, let M: u32>(
+pub unconstrained fn add_private_authwit_from_call<let M: u32, let N: u32, T>(
```
Aligned here the ordering of generics with PrivateCall. While not really necessary I was not sure if M and N is mixed up somewhere when I was debugging what turned out to be an unrelated issue.

benesjan commented on noir-projects/aztec-nr/aztec/src/test/helpers/test_environment.nr:
````diff
@@ -494,10 +494,10 @@ impl TestEnvironment {
     /// let return_value = env.call_private(caller, SampleContract::at(contract_addr).sample_private_function());
     /// ```
-    pub unconstrained fn call_private<T, let M: u32>(
+    pub unconstrained fn call_private<let M: u32, let N: u32, T>(
````
same here - aligned the ordering with PrivateCall

benesjan commented on noir-projects/aztec-nr/aztec/src/test/helpers/utils.nr:
```diff
@@ -9,13 +9,13 @@ use protocol_types::{
 use std::meta::derive;
 
-pub struct ContractDeployment<let N: u32> {
+pub struct ContractDeployment<let O: u32> {
```
Didn't want to overshadow the names of the generics in PrivateCall so I did the changes here. Also did this while debugging an issue.

dbanks12 reviewed (commented):
Public context and avm transpiler stuff looks good to me!

mverzilli commented on noir-projects/aztec-nr/aztec/src/context/calls.nr:
```diff
@@ -100,22 +100,22 @@ impl<let M: u32, T> PrivateStaticCall<M, T> {
     pub target_contract: AztecAddress,
     pub selector: FunctionSelector,
     pub name: str<M>,
     // TODO(F-131): Determine the length at comptime and drop the use of slices.
```
Shouldn't this comment go?

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/18417/commits/e3eb704b7a9d33fb4cb32508f33281935a8a5ee1

mverzilli commented on noir-projects/aztec-nr/aztec/src/context/public_context.nr:
```diff
@@ -1002,19 +1002,21 @@ unconstrained fn return_opcode(returndata: [Field]) {}
     da_gas_allocation: u32,
     address: AztecAddress,
-    args: [Field],
+    length: u32,
+    args: [Field; N],
 ) {}
```
Without context, `length` looks like a redundant parameter given args has `N` size. I guess there's a good reason for this so maybe it could be helpful to add a comment explaining

benesjan replied:
Good point. Addressed in https://github.com/AztecProtocol/aztec-packages/pull/18417/commits/e3eb704b7a9d33fb4cb32508f33281935a8a5ee1 and in https://github.com/AztecProtocol/aztec-packages/pull/18417/commits/e73ead2c19cde8640dd9e58cc7b4a9ffd5ef06ae

mverzilli commented on noir-projects/aztec-nr/aztec/src/test/helpers/authwit.nr:
```diff
@@ -10,11 +10,11 @@ use protocol_types::{
 };
 
-pub unconstrained fn add_private_authwit_from_call<T, let M: u32>(
+pub unconstrained fn add_private_authwit_from_call<let M: u32, let N: u32, T>(
```
Would be nice to have some comment briefly defining what M, N and T stand for. Or, if Noir allows, using more descriptive names for said type variables. eg: `add_private_authwit_from_call<let CallNameLen: u32, ArgsLen: u32, CallReturnType>`)

benesjan replied:
You have a point but this is all over the place and it's a bit unrelated so doesn't make sense to do in this PR.

mverzilli commented on noir-projects/aztec-nr/aztec/src/test/helpers/authwit.nr:
```diff
@@ -33,11 +33,11 @@ pub unconstrained fn add_private_authwit_from_call<T, let M: u32>(
 }
 
-pub unconstrained fn add_public_authwit_from_call<T, let M: u32>(
+pub unconstrained fn add_public_authwit_from_call<let M: u32, let N: u32, T>(
```
ditto

mverzilli reviewed (approved):
Nice! Only general comment is I would try to use more descriptive names for parametric types (especially when a function or type ends up having more than one, like many cases in this PR where there's M, N and T).

I don't know if Noir supports names with more than letter in such cases though, so feel free to ignore me

benesjan committed:
refactor: call args as an array instead of a slice

Usage of slices is now discouraged and for this reason I am dropping their use from calls.

This made me drop the `hash_args` function and rename `hash_args_array` to just `hash_args` as the original was no longer in use. Also `poseidon2_hash_with_separator_slice` got dropped due to lack of use.
